src/dependency_injector/providers.pyi

```diff
@@ -144,7 +145,7 @@ class DependenciesContainer(Object):
 class Callable(Provider[T]):
     def __init__(self, provides: Optional[_Callable[..., T]] = None, *args: Injection, **kwargs: Injection) -> None: ...
     @property
-    def provides(self) -> Optional[T]: ...
+    def provides(self) -> Type[T]: ...
```

Comment:
So in that case we better return ``_Callable[..., T]``. Attribute ``.provides`` should return what was provided in ``.__init__()`` or ``.set_provides()``. Could you, please, update it?

Reply:
Sure, I can. But If I could then propose here, because we can have the providers declared in two ways (they're semantically correct):

```python
provider10 = providers.Callable(Cat)
```

and this

```python
provider10 = providers.Callable[Animal](Cat)
```

mypy complains about the latter expecting that `Cat` (`Type[Cat]`) is not `_Callable[..., T]`.

what do you think of using `Union[Type[T], _Callable[..., T]]`?

I was going to propose that in another PR, but as you brought up this point, I thought it would be a pertinent question 😅

Reply:
Yep, that makes sense. Go for it

Reply:
Well, that's embarrassing... I added the tests using the explicit type on typevar

```
provider10 = providers.Callable[Animal](Cat)
```

and mypy didn't complain 😅 

I might have done something wrong on my project 🤔 

anyway, changed the code as you suggested at 47cf4c253b5470294e2cb7e534b2b118169c731b